### ISSUE TITLE
Sending multiple tabs via "Send tabs to device" results in strange relationships

### ISSUE THREAD

**Issue by irvinm** [piroor/treestyletab#2419]
**Environment**  

Windows 10 \ FF 70.0.1 \ TST 3.2.4

**Situation**  

I normally start the day checking my phone for interesting news or articles I might want to read.  Once you install Firefox on your phone, it installs a means to articles directly in Firefox or to send it to Firefox on another machine (Share -> Add to Firefox -> Send to other devices -> device name list)  Some articles I don't want to read on the phone or want to read later and I will send it to my desktop.

If I send multiple tabs, when the synchronization starts, it adds all of the tabs currently queued up under the currently selected tab.

**Problem**  

The behavior of how the tabs are created is not consistent.  All tabs are properly created but the relationship (child vs. sibling), the sequencing (are all 5 next to each other) is not consistent, and the "depth" of each tab can potentially be different.

**How to reproduce**

1. Send multiple items to be shared with Firefox using the "Send to other devices" method.  (I also tested this just using Firefox on 2 different machines by selecting multiple tabs from one Firefox instance and using the "Send tabs to device" right-click option on the tabs.)
2. On the "receiving" Firefox, just wait for the tabs to come across or manually trigger the synchronization by clicking your Firefox profile icon -> Sync now.

**Results**

![Snag_628b3b](https://user-images.githubusercontent.com/979729/68677934-f414f180-052a-11ea-830c-07d75bdb5653.png)

In this example, "Paid Services" is the currently selected tab and the 6 loaded tabs with the blue globes under the favicons were the tabs sent to my desktop.  Notice that the first 2 tabs were loaded under the paid services tab but are not actually related to that tab ... if I collapse "Paid Services" the first 2 tabs remain outside.  The third tab ("Exploring the ...") WAS created as a child of that tab.  The fourth tab ("Handling synonyms...") wasn't created next to the third as the discarded tab "WIRED" was between them.  The 5th and 6th tabs followed a similar pattern.

![Snag_62aa0e](https://user-images.githubusercontent.com/979729/68677948-f8410f00-052a-11ea-8efa-b595923ac2e9.png)

I wanted to try some other scenarios and in this one (different starting tab with different tabs being sent across)  The pattern looks to be the same.

![Snag_6ddd7d](https://user-images.githubusercontent.com/979729/68678821-ae592880-052c-11ea-89d1-1e71c736b670.png)

In this example, I sent the same 5 tabs across but used a different active tab.  Here it creates the first 3 as siblings, skips over one discarded tab before creating the 4th.  The last is added as a child of an existing tab.

**Expected Results**

I would expect any number of tabs to be sent across and ALL to be placed consecutively either as siblings or children of the currently selected tab.  This behavior has been around for quite some time, I just found some time to finally write it up.  This should not be related to any recently FF or TST changes.  I would also expect there is a good chance this might be a "bug of Firefox itself".

**Comment by irvinm** [comment]
@piroor ... one other side affect of this is that in the first example with the tabs under "Paid Services", if I leave things just as they are and restart Firefox, all of the tabs lose the "Paid Services" tab as their parent and they all become siblings.  I was able to reproduce this affect twice in a row. (Below is my 2nd attempt and after restarting Firefox)

![Snag_e28c27](https://user-images.githubusercontent.com/979729/68689523-7bb82b80-053e-11ea-8a61-76cb6ee5003e.png)

**Comment by piroor** [comment]
Received tabs are opened as tabs like opened from bookmarks or history entries. Thus "how opened tabs are treated" depends on options of Firefox, TST, and other addons. (For example, on my environment they are opened at the end of the tab bar as independent tabs, not children of the active tab.)

I've tried to determine a tab is sent from another device or not, but sadly there is no hint...

**Comment by irvinm** [comment]
@piroor I spent some more time on this and have more information and it all seems related to browser.tabs.insertAfterCurrent = true.  When this value is false, I see the same behavior as you where the tabs are created sequentially (at the same level) at the end of the tab list.

In a clean profile with only TST installed, browser.tabs.insertAfterCurrent = true, sending over 5 tabs, and the active tab is:

... the last tab - all 5 tabs are created in sequence and at the same level.  (Good)
![Snag_5f20f7](https://user-images.githubusercontent.com/979729/68863125-bc3db380-06bc-11ea-82fe-510f31b42229.png)

... next to last tab - tabs are created in sequence and at the same level.  (Good)
![Snag_622fcd](https://user-images.githubusercontent.com/979729/68863352-33734780-06bd-11ea-8ff2-636f6418a24e.png)

... tab in the middle of a tree - first 3 tabs are created in sequence, but last 2 have existing tabs in between.  (Bad)
![Snag_64d3c4](https://user-images.githubusercontent.com/979729/68863555-9a90fc00-06bd-11ea-97c7-6a496fbc9794.png)

... root tab of a tree - I see a mix of indentation of the tabs. (Bad)
![Snag_66be2f](https://user-images.githubusercontent.com/979729/68863717-e5ab0f00-06bd-11ea-8382-bc691961ede8.png)

... in addition, restarting Firefox at this point results in the tree losing its relationships.
![Snag_680e00](https://user-images.githubusercontent.com/979729/68863845-18ed9e00-06be-11ea-9c03-f2f45cb7e161.png)

... root tab of a tree that isn't group-tab.html - again tab sequencing and relationships are wrong. (bad)
![Snag_6c1b89](https://user-images.githubusercontent.com/979729/68864201-b6e16880-06be-11ea-81e6-10a9d6f58f7d.png)

... tab above a tree - again first 3 are created properly, but the next 2 created with tabs in between which are part of the tree below it.
![Snag_744ab0](https://user-images.githubusercontent.com/979729/68864949-f6f51b00-06bf-11ea-9cb6-5c4c2308cf9a.png)


@piroor this is pretty easy for me to try different things if you want me to.  I just launch Firefox, then launch another Firefox (different profile) using the -no-remote option and just try different scenarios.  No phone\tablet are needed.

**Comment by piroor** [comment]
Thanks! I've successfully reproduced this problem on my desktop PC. I'm using both Firefox 70 and Nightly 72.0a1, and this happens when I send multiple tabs from one to another with the setting. (I've confirmed that Firefox Android is unnecessary for debugging.)

**Comment by piroor** [comment]
And after more trying I've realized that Sync is also unnecessary to reproduce the problem. I've successfully reproduced this problem just a script in the console: `for (let i = 0; i < 3; i++) { openUILinkIn(`about:blank#${i}`, 'tabshifted', { triggeringPrincipal: Services.scriptSecurityManager.getSystemPrincipal() }); }` This means that this problem may happen on any Firefox feature opening multiple tabs synchronously...

**Comment by piroor** [comment]
The commit 817bf64 partially fixes this issue: now tabs opened in the middle of an existing tree are correctly attached based on their position. However still there is another problem: received tabs are not opened successively. If you send 5 or more tabs they are received as gapped tabs.

**Comment by piroor** [comment]
Finally I've understood why tabs are opened with gap. I think this is an inconsistency problem of webExtensions API itself.

I tested a simple `tabs.onCreated` handler and multiple received tabs, without TST. This screenshot indicates what was happen:
![image](https://user-images.githubusercontent.com/70062/68946901-6d486900-07f7-11ea-8054-9e4654a61617.png)
Please look at their `id` and `index`. They are identified sequentially but index of last four tabs are reversed. This means that their `index` property doesn't mean the index when they are opened. So, for example, when the tab `29` is notified:

* Originally there are existing tabs: `[3, 2, 27, 5, 10, 11, 12, 13, 14, 15]`
  * `5` is a parent and the active tab.
  * `10`-`14` are children of `5`.
* `28` is notified with index `4`, so it is inserted to the position: `[3, 2, 27, 5, 28, 10, 11, 12, 13, 14, 15]`
  * `28` is inserted between `5` (parent) and `10` (child).
  * Then TST attaches `28` to `5` as its new first child.
* `29` is notified with index `7`, so it is insreted to the position: `[3, 2, 27, 5, 28, 10, 11, 29, 12, 13, 14, 15]`
  * Please note it should be inserted as `[3, 2, 27, 5, 29, 28, 10, 11, 12, 13, 14, 15]` because it is opened as the next to the active tab `5` - its expected index is `4` not `7` on this timing.

Currently I have no good idea to solve this problem on TST side. Buffering and re-indexing looks to be a solution, but I cannot determine which tabs are opened at same time.

Thus I think that Firefox should notify `5` as the index of the tab `29` on such cases. In other words this should be tracked as a bug of  WebExtensions API and should be fixed on Firefox side...

**Comment by irvinm** [comment]
@piroor thank you for the partial fix and investigating this situation.  It definitely looks like an issue on the Firefox side.  Did you plan on submitting a defect to them on this?

**Comment by piroor** [comment]
I'm planning to file a bug to bugzilla, and now researching a minimum steps to reproduce.

**Comment by piroor** [comment]
Finally I've understood that [Sync opens received tabs](https://searchfox.org/mozilla-central/rev/cac340f10816707e91c46db6d285f80259420f07/browser/components/BrowserGlue.jsm#3524) like this:

```javascript
(async () => {
  const uris = [];
  for (let i = 0; i < 5; i++) {
    uris.push(`about:blank#${i}`);
  }
  const openTab = async uri => {
    gBrowser.addWebTab(uri);
  };
  const firstTab = await openTab(uris[0]);
  uris.slice(1).map(openTab);
})()
```

**Comment by piroor** [comment]
More short version:

```javascript
(async () => {
  await gBrowser.addWebTab(`about:blank#first`)
  for (let i = 0; i < 5; i++) {
    gBrowser.addWebTab(`about:blank#${i}`);
  }
})()
```

The most important point to reproduce is: we need to do `await` for the first tab.

**Comment by piroor** [comment]
I've filed: [1596787 - tabs.onCreated callback receives tabs.Tab object with wrong index when multiple tabs are opened via "Send Tab to Device" with "browser.tabs.insertAfterCurrent"="true"](https://bugzilla.mozilla.org/show_bug.cgi?id=1596787)

**Comment by irvinm** [comment]
Thank you sir.  Your effort and knowledge is appreciated and amazing. (I've added my vote to the issue and will watch its progress)

**Comment by irvinm** [comment]
@piroor I see you were marked the "assignee" for the Bugzilla defect 2 months ago.  Are you continuing to work on this one on the Firefox side?  (I don't think so, but thought I would ask)

**Comment by piroor** [comment]
Sadly my approach is stuck in mud on bugzilla. Currently I have no good idea to solve concerns presented by the reviewer and I need help from tab and WebExtensions expert...

**Comment by irvinm** [comment]
@piroor Did you want to leave a similar comment on Bugzilla and unassign yourself from the bug?  (That would at least indicate to any triage team that no one is progressing the issue)  It is obviously still an issue, but affects very few people.

**Comment by piroor** [comment]
Because the bug https://bugzilla.mozilla.org/show_bug.cgi?id=1417183 (about genuine WebExtensions API proposal around Firefox Sync) became WONTFIX, now I'm trying to simulate "Send Tab to Device" features based on the sync storage. It allows us to send multiple tabs including their tree structure. I think the change will make this issue obsolete.

**Comment by irvinm** [comment]
@piroor I saw your work on this based on the bug feedback.  Keep in mind, this issue specifically is when tabs are sent to a machine with TST on it and how TST handles receiving them.  (Although I love that TST will be able to emulate initiating sending tabs as well)

I have actually been meaning to update this issue with some additional details, but it hasn't been a big issue as I understand tabs I send from my mobile to my PC need to be handled differently.  For example, I believe if multiple tabs are sent from my mobile while the PC has not yet started, then when the PC starts and Firefox loads that the behavior of the inserted tabs is different than if the PC and Firefox were already running.  (If sent before Firefox is running, I believe they are all inserted at the root indentation level, but already running they are inserted at the proper indentation)

The 2 biggest issues are:

- How the tabs are inserted into a tree.  (Both indentation and possibly tab order)
- How some inserted tabs can cause the tree into which it was inserted to become "flatten" after Firefox restarts. (I haven't seen this often, but I think it is possible when the tab is inserted at the root indentation and I don't close or move the tab [thus re-establishing its relationship in the tree] and restart Firefox as the tree from that tab down are all then at the same level)

I'll see if I can get more current examples written up to see if there is anything that can be done.  It may not be worth the effort.

**Comment by irvinm** [comment]
@piroor I know we are a bit stuck on this problem, but wanted you to know there is a related issue to this.

Even if I send a small number of tabs (say 1-2), TST handles receiving them differently depending if Firefox is currently running or not.

With Firefox already running, sending 2 tabs are handled properly and inserted into the tree as expected.

![image](https://user-images.githubusercontent.com/979729/140317159-e52dd74a-f55d-49f4-b4ef-976f7e136b47.png)

If Firefox is not running when the tabs are sent then started up, the tabs are not inserted properly into the tree.  (Example, reading an article on my phone in Firefox, then sending to my desktop for reading later when I turn on my machine)

![image](https://user-images.githubusercontent.com/979729/140317268-b22fba67-2a1c-4a4c-80e3-a1dd516c6fa1.png)

**Comment by piroor** [comment]
@irvinm I think the "wrongly inserted received tabs on startup" problem should be tracked separately from this. Could you open another issue?

**Comment by piroor** [comment]
Hmm, should I create a new label to indicate the state "unsolved but closed just to prevent in-progress issues are buried in too many staled issues"?

**Comment by irvinm** [comment]
How about a label like "keep open" that you can use as a generic one that you don't want to be automatically closed?  (I think the bot is working very well in general and has closed several items that should be closed (people forget to just close it) or just don't have enough information but they don't respond back)

**Comment by irvinm** [comment]
One other thought on this ... is there a way to have Github place a default label on all new items?  (Like "needs triage")  This would give you a way to do a search for any items that you have not initially reviewed.  Once you have reviewed, you could remove that label and apply any others that are appropriate.  I currently see there are 2 items more than a week old in this state, but when sorted by "recently updated" they are moving down the list.

**Comment by irvinm** [comment]
@piroor any thoughts on the labels and\or reopening this item?

**Comment by piroor** [comment]
Basically

**Comment by piroor** [comment]
I've reopened this because I've started to plan workaround about  this problem on TST side.

Basically I think that issues on this issue tracker must be opened or closed based on the state on this TST project. If we can do something about it or something in-progress tasks around the issue, it should be open. Otherwise it should be closed.

With this policy, not yet triaged issues should be open. I'm OK to set "not triaged" label for such issues, but currently I don't know how to do that without the bot. Issue Form may be the solution.

----

When I send multiple tabs from a device to another, they are opened in the destination device with these indices:

(X = the index of the active tab, N = the number of received tabs)

* X + 1
* X + 1 + (N - 2)
* X + 1 + (N - 3)
* ...
* X + 1 + (N - N)

So I think we can fixup the position of opened tabs if they are opened with indices in this pattern.

**Comment by piroor** [comment]
Indices of received tabs when 10 tabs are sent:

|A) When there is no following tab|B) When there only one tab following to the active tab|C) When there are 2 tabs following to the active tab|D) When there are 3 tabs following to the active tab|E) When there are 4 tabs following to the active tab|F) When there are 5 tabs following to the active tab|G) When there are 6 tabs following to the active tab|H) When there are 7 tabs following to the active tab|I) When there are 8 tabs following to the active tab|J) When there are 9 tabs following to the active tab|
|---|---|---|---|---|---|---|---|---|---|
|1|1|1|1|1|1|1|1|1|1|
|2|3|4|5|6|7|8|9|9|9|
|3|4|5|6|7|8|8|8|8|8|
|4|5|6|7|7|7|7|7|7|7|
|5|6|6|6|6|6|6|6|6|6|
|5|5|5|5|5|5|5|5|5|5|
|4|4|4|4|4|4|4|4|4|4|
|3|3|3|3|3|3|3|3|3|3|
|2|2|2|2|2|2|2|2|2|2|
|1|1|1|1|1|1|1|1|1|1|

**Comment by piroor** [comment]
Experimental script to calculate indices to detect the pattern:

```javascript
for (let activeIndex = 0; activeIndex < 3; activeIndex++) {
  for (let followingTabsCount = 0; followingTabsCount < 15; followingTabsCount++) {
    const createdCount = 10;
    const indices = [activeIndex + 1];

    const overTabsCount = Math.max(0, createdCount - followingTabsCount);
    const shouldCountDown = Math.min(createdCount - 1, createdCount - Math.floor(overTabsCount / 2));
    const shouldCountUp = createdCount - shouldCountDown - 1;
    for (let i = 0; i < shouldCountUp; i++) {
      indices.push(activeIndex + 2 + i + (createdCount - overTabsCount));
    }
    for (let i = shouldCountDown - 1; i > -1; i--) {
      indices.push(activeIndex + 1 + i);
    }
    console.log({ overTabsCount, shouldCountUp, shouldCountDown, size: indices.length }, indices.join(','));
  }
}
```

**Comment by piroor** [comment]
I've implemented a workaround with the commit 19194dc. When opened tabs are detected as "sent from other device with `browser.tabs.insertAfterCurrent`=`true`" based on their index, TST tries to put the gather of tabs to near the first sent tab. Please remind that the order of tabs are still mixed due to Firefox's bug or design (and we have no chance to know their original order).
